### DISCUSSION TITLE
TCP Healthcheck

### DISCUSSION THREAD

**Discussion by Kinsiinoo** [qdm12/gluetun#929] [Q&A] [answered]
Hi, so I upgraded from [v3.26.0](https://github.com/qdm12/gluetun/releases/tag/v3.26.0) to c33fb8b (latest docker image that time)  and I saw these #920 , #906.
The new TCP Healthcheck is a little bit unreliable with `github.com:443` and I get this log entry 1-2 times per day since the update:
```
2022-04-05T01:10:57Z INFO [healthcheck] unhealthy: cannot dial: dial tcp4 140.82.121.4:443: i/o timeout
2022-04-05T01:11:02Z INFO [healthcheck] program has been unhealthy for 5s: restarting VPN
```
It is usually healthy again after that small outage but disconnects the network from the containers connected to Gluetun, which then do not recover and I need to restart manually. (also disconnects ports that are passed due to containers `[firewall] removing allowed port XXXXX`)

**My questions are:**
1. Does anybody know an url which is more stable than `github.com:443`? Because if I'm correct we can change that with `HEALTH_TARGET_ADDRESS` config.
2. If not then can I change back the docker image version **without trouble** and which is the most recommended? (I haven't rewritten the variable names to their new names yet)

Thanks in advance!

**Comment by qdm12** [comment, marked as answer]
1. You can try `1.1.1.1:443` let me know if it works better
2. Yes you should be able. At least back to v3.26.0 it should work. Note the health system was broken though, so it won't detect a connection drop.

> It is usually healthy again after that small outage but disconnects the network from the containers connected to Gluetun, which then do not recover and I need to restart manually.

That shouldn't happen as long as the container isn't retarted. Do you have any other program restarting unhealthy containers? Or do your containers lose connection although the gluetun container didn't restart?

> also disconnects ports that are passed due to containers

Can you share the logs with `FIREWALL_DEBUG=on` when this happens? That may be a bug if it doesn't re-establish allowed ports on an internal vpn reboot.

**Reply by Kinsiinoo** [reply]
>You can try 1.1.1.1:443 let me know if it works better

I pinged the default github address from two computers (two different locations and networks) for 24 hours and it was 1-2 packet loss.
Couldn't that cause the thing I mentioned?

I'll try `1.1.1.1:443` and give feedback.

>That shouldn't happen as long as the container isn't retarted. Do you have any other program restarting unhealthy containers? Or do your containers lose connection although the gluetun container didn't restart?

>Can you share the logs with `FIREWALL_DEBUG=on` when this happens? That may be a bug if it doesn't re-establish allowed ports on an internal vpn reboot.

I don't have any program that restart containers. (I only use portainer)

One of the container connected to Gluetun is a qBit client (this is the container that is problematic even after the connection returns), after Gluetun reestabilish connection, this client connects, but since the port has already been dropped (ignoring that it has since been re-enabled), it changes to a state that only fixable with a manual restart.
So this is a killswitch which would be a good thing that if the VPN is interrupted then the traffic will not go through the insecure internet. But the qBit client not be able to recover from it and it will be "killed" by Gluetun too often due to the momentary drop of Github connection.

**Reply by qdm12** [reply]
> So this is a killswitch which would be a good thing that if the VPN is interrupted then the traffic will not go through the insecure internet.

This is already in place.

> But the qBit client not be able to recover from it and it will be "killed" by Gluetun too often due to the momentary drop of Github connection.

What do you mean? Is it completely offline (try `docker exec -it qbitorrent wget -qO- https://google.com` for example), or is it just the VPN forwarded port that is not re-setup properly?

**Reply by Kinsiinoo** [reply]
> This is already in place.

Yeah I know, I just mentioned it is a good thing, just too sensitive with `github.com:443`. 🧐 

> What do you mean? Is it completely offline (try `docker exec -it qbitorrent wget -qO- https://google.com` for example), or is it just the VPN forwarded port that is not re-setup properly?

Yeah, it is just the VPN forwarded port that is not re-setup properly but that can be a problem with qBit or qBit container image, I don't know. Not 100% Gluetun bug.

**I marked your answer as a solution, I'm using `1.1.1.1:443` ever since and it is 1000x more stable than `github.com:443`.
I've been pinging it all the time since your answer. There were not a single packet loss during this time!
Thanks!** 👍🏻 😎

**Reply by qdm12** [reply]
I changed it to `cloudflare.com:443` as default in 10a13bc8a7498028ed67f16a292c62aa3e3c1def such that it also checks the DNS is working, since `1.1.1.1:443` doesn't test the DNS.

**Reply by iluvatyr** [reply]
just wanted to note that it removed the forwarded port here too. Since my connection was very unstable due to the wireguard server being unstable, it had to add and remove the port several times. In the end, I tested with a port checker tool and the port was not forwarded anymore. A manual restart of the docker-compose solved this. During all this time, both gluetun and the qbittorrent client connected to it were still online and working, just the port forward got lost.
``` 
2023-04-20T06:44:56+02:00 INFO [firewall] removing allowed port XXXX...
2023-04-20T06:44:56+02:00 INFO [vpn] starting
2023-04-20T06:44:56+02:00 INFO [firewall] allowing VPN connection...
2023-04-20T06:44:56+02:00 INFO [wireguard] Using available kernelspace implementation
2023-04-20T06:44:56+02:00 INFO [wireguard] Connecting to XXXXX:XXX
2023-04-20T06:44:56+02:00 INFO [wireguard] Wireguard is up
2023-04-20T06:44:56+02:00 INFO [firewall] setting allowed input port XXXX through interface tun0...
2023-04-20T06:44:59+02:00 ERROR [ip getter] Get "https://ipinfo.io/": dial tcp: lookup ipinfo.io on 10.13.13.1:53: read udp 10.13.13.2:49360->10.13.13.1:53: i/o timeout
2023-04-20T06:44:59+02:00 INFO [ip getter] retrying in 5s
2023-04-20T06:45:10+02:00 INFO [ip getter] Public IP address is XXXXXXXX
2023-04-20T06:45:11+02:00 INFO [healthcheck] healthy!
2023-04-20T06:45:24+02:00 INFO [healthcheck] unhealthy: dialing: dial tcp4: lookup github.com: i/o timeout
2023-04-20T06:45:26+02:00 INFO [healthcheck] healthy!
```

@qdm12 Is switching to cloudflare.com as a healthcheck URL fixing this or was this another problem, since I dont get how changing the URL would fix this
